R/splitBy.R

```diff
@@ -0,0 +1,264 @@
+#' Split \code{SummarizedExperiment} column-wise or row-wise based on grouping variable
+#'
+#' @param x A
+#'   \code{\link[SummarizedExperiment:SummarizedExperiment-class]{SummarizedExperiment}}
+#'   object
+#'
+#' @param grouping A single character value for selecting the grouping variable
+#'   from \code{colData} or \code{rowData}.
+#' 
+#' @param MARGIN A single numeric value, 1 (row) or 2 (col),  for selecting from 
+#'   where grouping variable should be searched.
+#'   
+#' @param ... 
+#' \itemize{
+#'   \item{\code{use_names} A boolean value to select whether to name elements of
+#'   list by their group names.}
+#' }
+#'
+#' @return
+#' List of \code{SummarizedExperiment} objects in \code{SimpleList} format.
+#'
+#'
+#' @details
+#' \code{splitBy} split data based on grouping variable. Splitting can be done
+#' column-wise or row-wise. Returned value is a list of \code{SummarizedExperiment}
+#' objects; each element containing members of each group.
+#'
+#' @seealso
+#' \code{\link[=splitByRanks]{splitByRanks}}
+#' \code{\link[=unsplitByRanks]{unsplitByRanks}}
+#' \code{\link[=merge-methods]{mergeRows}},
+#' \code{\link[scuttle:sumCountsAcrossFeatures]{sumCountsAcrossFeatures}},
+#' \code{\link[=agglomerate-methods]{agglomerateByRank}},
+#' \code{\link[SingleCellExperiment:altExps]{altExps}},
+#' \code{\link[SingleCellExperiment:splitAltExps]{splitAltExps}}
+#'
+#'
+#' @name splitBy
+#' @export
+#' 
+#' @author Leo Lahti and Tuomas Borman. Contact: \url{microbiome.github.io}
+#'
+#' @examples
+#' data(GlobalPatterns)
+#' tse <- GlobalPatterns
+#' # Split data based on SampleType. 
+#' se_list <- splitBy(tse, grouping = "SampleType")
+#' 
+#' se_list
+#' 
+#' # Create arbitrary groups
+#' colData(tse)$group <- sample(1:10, ncol(tse), replace = TRUE)
+#' rowData(tse)$group <- sample(1:10, nrow(tse), replace = TRUE)
+#' 
+#' # If variable named equally can be found from both colData and rowData, 
+#' # MARGIN must be specified
+#' se_list <- splitBy(tse, grouping = "group", MARGIN = 2)
+#' 
+#' # It is possible to split data also in row-wise
+#' se_list <- splitBy(tse, grouping = "group", MARGIN = 1)
+#' 
+#' # Split data based on Phyla
+#' se_list <- splitBy(tse, "Phylum")
+#' 
+#' se_list
+#' 
+#' # List of SE objects is returned. 
+#' # Each element is named based on their group name. If you don't want to name
+#' # elements, use use_name = FALSE
+#' se_list <- splitBy(tse, grouping = "SampleType", use_name = FALSE)
+#' 
+#' # If you want to combine groups back together, you can use unsplitBy
+#' unsplitBy(se_list)
+#' 
+NULL
+
+#' @rdname splitBy
+#' @export
+setGeneric("splitBy",
+           signature = "x",
+           function(x, grouping = NULL, MARGIN = NULL, ...)
+               standardGeneric("splitBy"))
+
+#' @rdname splitBy
+#' @export
+setMethod("splitBy", signature = c(x = "ANY"),
```

Comment:
The scope of the function is to broad. To be compatible with the expectations from `splitByRank`, I think you need to mirror the S4 methods for `SE`, `SCE` and `TSE`. This is very important to allow future developments, e.g. tree clustering and such. Only with those interfaces will this be possible.